relay-metrics/src/aggregator/inner.rs

```diff
@@ -0,0 +1,926 @@
+        // but still shrinking a little bit so that over time we free up memory again
+        // without re-allocating too many times.
+        //
+        // Also re-use the original hasher, no need to create a new/expensive one.
+        self.slots.push_back(Slot {
+            buckets: HashMap::with_capacity_and_hasher(
+                slot.buckets.len(),
+                slot.buckets.hasher().clone(),
+            ),
+            ..slot
+        });
+
+        // Increment to the next partition.
+        //
+        // Effectively advance time by `bucket_interval / num_partitions`.
+        self.head += 1;
+
+        Partition {
+            partition_key,
+            buckets: slot.buckets,
+            stats,
+        }
+    }
+
+    /// Consumes the aggregator and returns an iterator over all contained partitions.
+    pub fn into_partitions(self) -> impl Iterator<Item = Partition> {
+        self.slots.into_iter().map(|slot| Partition {
+            partition_key: slot.partition_key,
+            buckets: slot.buckets,
+            stats: PartitionStats::from(&slot.stats),
+        })
+    }
+}
+
+impl fmt::Debug for Inner {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut list = f.debug_list();
+        list.entry(&self.stats);
+        for (i, v) in self.slots.iter().enumerate() {
+            let head_partitions = self.head % u64::from(self.num_partitions);
+            let time_offset = (head_partitions + i as u64) / u64::from(self.num_partitions);
+            let head_time = self.head / u64::from(self.num_partitions);
+            let time = (head_time + time_offset) * u64::from(self.bucket_interval);
+            match v.is_empty() {
+                // Make the output shorter with a string until `entry_with` is stable.
+                true => list.entry(&format!("({time}, {v:?})")),
+                false => list.entry(&(time, v)),
+            };
+        }
+        list.finish()
+    }
+}
+
+#[derive(Default)]
+struct Slot {
+    pub partition_key: u32,
+    pub stats: stats::Slot,
+    pub buckets: HashMap<BucketKey, BucketData>,
+}
+
+impl Slot {
+    fn is_empty(&self) -> bool {
+        self.stats == Default::default() && self.buckets.is_empty()
+    }
+}
+
+impl fmt::Debug for Slot {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if self.is_empty() {
+            write!(f, "Slot({})", self.partition_key)
+        } else {
+            #[cfg(test)]
+            let buckets = &self.buckets.iter().collect::<BTreeMap<_, _>>();
+            #[cfg(not(test))]
+            let buckets = &self.buckets;
+
+            f.debug_struct("Slot")
+                .field("partition_key", &self.partition_key)
+                .field("stats", &self.stats)
+                .field("buckets", buckets)
+                .finish()
+        }
+    }
+}
+
+struct SlotDiff {
+    min: u64,
+    max: u64,
+}
```

Comment:
Took me a while to understand this type, maybe naming helps
```suggestion
struct RelativeRange {
    max_secs_in_past: u64,
    max_secs_in_future: u64,
}
```